XRDXRFutils/gaussnewton_multi.py

```diff
@@ -0,0 +1,389 @@
+        return self.synthetic_spectrum_multiphase(False).T
+
+    def z_multiphase(self):
+        """Synthetic spectrum of each phase."""
+        return self.synthetic_spectrum_multiphase(True).T
+
+
+    def z0(self):
+        """Synthetic spectrum of the combination of phases, with gamma = 1 for all peaks."""
+        return self.synthetic_spectrum_multiphase(False).sum(axis = 1)
+
+    def z(self):
+        """Synthetic spectrum of the combination of phases."""
+        return self.synthetic_spectrum_multiphase(True).sum(axis = 1)
+
+
+    ### Calculations for fit ###
+
+    def calculate_components(self):
+        mu = [arr[newaxis, :] for arr in self.mu]
+        I = [arr[newaxis, :] for arr in self.I]
+        gamma = [arr[newaxis, :] for arr in self.gamma]
+        sigma2 = [arr[newaxis, :] for arr in self.sigma2]
+        theta = [arr[:, newaxis] for arr in self.theta]
+
+        self.component_core = [exp((theta[idx] - mu[idx])**2 / (-2 * sigma2[idx])) for idx in range(self.n_phases)]
+        self.component_full = [I[idx] * gamma[idx] * self.component_core[idx] for idx in range(self.n_phases)]
+
+    def del_components(self):
+        del self.component_core
+        del self.component_full
+
+
+    def der_f_a_s_beta(self):
+        mu = [arr[newaxis, :] for arr in self.mu]
+        sigma2 = [arr[newaxis, :] for arr in self.sigma2]
+        channel = self.channel[:, newaxis]
+        theta = [arr[:, newaxis] for arr in self.theta]
+
+        a = self.opt[:self.n_phases]
+        s = self.opt[self.n_phases]
+
+        der_denominator = [(channel + a[idx])**2 + s**2 for idx in range(self.n_phases)]
+        der_theta_a = [rad2deg(s / der_denominator[idx]) for idx in range(self.n_phases)]
+        der_theta_s = [- rad2deg((channel + a[idx]) / der_denominator[idx]) for idx in range(self.n_phases)]
+
+        aux = [(self.component_full[idx] * (theta[idx] - mu[idx]) / sigma2[idx]).sum(axis = 1, keepdims = True) for idx in range(self.n_phases)]
+
+        der_f_a = [- der_theta_a[idx] * aux[idx] for idx in range(self.n_phases)]
+        der_f_s = - concatenate([der_theta_s[idx] * aux[idx] for idx in range(self.n_phases)], axis = 1).sum(axis = 1, keepdims = True)
+        der_f_beta = - concatenate(aux, axis = 1).sum(axis = 1, keepdims = True)
+
+        return der_f_a + [der_f_s, der_f_beta]
+
+
+    def der_f_g(self):
+        I = [arr[newaxis, :] for arr in self.I]
+        return [I[idx] * self.component_core[idx] * self.der_w(self.g[idx]) for idx in range(self.n_phases)]
+
+
+    def der_f_tau(self):
+        mu = [arr[newaxis, :] for arr in self.mu]
+        sigma2 = [arr[newaxis, :] for arr in self.sigma2]
+        theta = [arr[:, newaxis] for arr in self.theta]
+        return [self.component_full[idx] * ((theta[idx] - mu[idx])**2 / (2 * sigma2[idx]**2)) * self.der_u(self.tau[idx]) for idx in range(self.n_phases)]
+
+
+    def fit(self, a = False, s = False, beta = False, gamma = False, sigma = False, alpha = 1, downsample = None):
```

Comment:
You should have a separated into a1 and a2 here. They are equivalent to s, beta and gamma. Currently, I can not fix one phase and iterate only the other one If I need to or at least I do not see how.

Reply:
Yes, I see your point. I'm going to deal with it.

Reply:
ok thanks

Reply:
You can test the new joined calibration in [this notebook](https://github.com/DarioZappala/XRDAnalysis/blob/master/XRD_1p_test_phase_shift__joined_fit.ipynb).

Reply:
In the notebook:
Calculate the shift in microns along the direction of the incident ray. Apply the formula from https://codimd.infn.it/TGkIwbqqTEqHft9AmigaEw# and then calculate the new z in microns.

Reply:
Done. Now I calculate Delta y (along the incident direction) and Delta z (along the direction orthogonal to the sample surface).
I've also updated in the same way the two notebooks that search for stratigraphy in a map with separate calibration ([here](https://github.com/DarioZappala/XRDAnalysis/blob/master/XRD_2d_phase_shift__separate_fit.ipynb)) and with joined calibration ([here](https://github.com/DarioZappala/XRDAnalysis/blob/master/XRD_2d_phase_shift__joined_fit.ipynb)).